### PR TITLE
Add decimal scaling to stable curve

### PR DESCRIPTION
**Pull request by elliotkennedy**
## Overview

- Add `token_a_factor` and `token_b_factor` fields to the `StableCurve` account
- Before any swap operation (or asymmetric deposit/withdrawal) - scale each input by the factor
- Such that:
```md
1 USDH = 1.000_000     = 1_000_000     => multiply by 10**3 = 1_000_000_000
1 UXD  = 1.000_000_000 = 1_000_000_000 => multiply by 10**0 = 1_000_000_000
```
- When scaling back down, each number should be rounded to favour the swap pool - round up amounts sent to the pool and round down amounts sent out of the pool
- Most changes are in `src/curve/stable.rs`

### CODE REVIEW

File: programs/hyperplane/src/curve/stable.rs [hunk 1]
```diff
@@ -216,14 +216,115 @@ fn compute_y(ann: u64, x: u128, d: u128) -> Result<u128> {
...
+    } else {
+        source_amount
+    };
+    Ok(amount)
+}
+
+fn scale_down(source_amount: u128, factor: u64, round_up: bool) -> Result<u128> {
+    let amount = if factor > 1 {
+        let amount = try_math!(source_amount.try_div(factor as u128))?;
+        if round_up && source_amount % factor as u128 > 0 {
```
Comment (elliotkennedy) [inline programs/hyperplane/src/curve/stable.rs:231]: Any suggestions to avoid the modulo + divide here?

Comment (oeble) [reply programs/hyperplane/src/curve/stable.rs]: Sadly, the "right" API is not available in stable rust. But anyway, if your computation is clearly written, the compiler will optimise it if the target offers relevant instructions (probably not the case for sbf/bpf).

To make the compiler job easier, you can avoid here-checked math since the only failing case is where `factor == 0` (which you don't handle btw). Also, you can replace the `%` with a simple multiplication (way less CU intensive):

```rust
let amount = source_amount / factor;
if amount * factor < source_amount {
    amount + 1;
} else {
    amount
}
```

Two extra points:
- In your current implementation, you compute `try_math!(source_amount.try_div(factor))?` twice which is not needed.
- `u128` division still costs way more CU than `U128` AFAIK so we should prefer the latter.

File: programs/hyperplane/src/curve/stable.rs [hunk 2]
```diff
@@ -216,14 +216,116 @@ fn compute_y(ann: u64, x: u128, d: u128) -> Result<u128> {
...
+    Ok(scaled)
+}
+
+pub fn scale_swap_outputs(
+    curve: &StableCurve,
+    new_destination_amount: u128,
+    trade_direction: TradeDirection,
+) -> Result<u128> {
+    let new_destination_amount = match trade_direction {
+        // round up to ensure the pool is favoured
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/hyperplane/src/curve/stable.rs]: Aha - sorry this should be clearer - I have now renamed `new_destination_amount` -> `new_pool_destination_amount` to clarify that this is the amount of destination token remaining in the pool after the swap.

You can see here that the bigger the `new_destination_amount`, the less the user will receive (`amount_swapped`), so rounding up is favouring the pool.

```rs
let amount_swapped = try_math!(pool_destination_amount.try_sub(scale_swap_outputs(
    self,
    new_destination_amount,
    trade_direction
)?))?;
```

File: programs/hyperplane/src/curve/stable.rs [hunk 3]
```diff
@@ -48,7 +46,7 @@ fn try_u8_power(a: &U256, b: u8) -> Result<U256> {
 fn try_u8_mul(a: &U256, b: u8) -> Result<U256> {
     let mut result = *a;
     for _ in 1..b {
-        result = result.try_add(*a)?;
+        result = try_math!(result.try_add(*a))?;
     }
     Ok(result)
 }
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/hyperplane/src/curve/stable.rs:52]: Yeah I'm not sure on either point but will look into improving in the future. I will be able to check CU with new tests in a future PR

File: programs/hyperplane/sim/src/lib.rs [hunk 4]
```diff
@@ -78,7 +79,11 @@ impl StableSwapModel {
         let mut d = s.clone();
         let ann = &self.amp_factor * &self.n_coins;
 
+        let mut iterations = 0;
         while d.abs_diff(&d_prev) > BigInt::one() {
+            if iterations >= MAX_ITERATIONS {
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/hyperplane/sim/src/lib.rs]: Nice - done

File: programs/hyperplane/src/curve/stable.rs [hunk 5]
```diff
@@ -216,14 +216,116 @@ fn compute_y(ann: u64, x: u128, d: u128) -> Result<u128> {
...
+            curve.token_b_factor,
+            true
+        ))?,
+        // round up to ensure the pool is favoured
+        TradeDirection::BtoA => try_math!(scale_down(
+            new_destination_amount,
+            curve.token_a_factor,
+            true
+        ))?,
+    };
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/hyperplane/src/curve/stable.rs]: Sounds good - done

File: programs/hyperplane/src/instructions/test/test_initialize_pool.rs [hunk 6]
```diff
@@ -612,6 +618,88 @@ fn test_initialize(
         accounts.initialize_pool().unwrap();
     }
 
+    // create invalid stable swap
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/hyperplane/src/instructions/test/test_initialize_pool.rs:621]: Ah it's just a hangover from the pre-anchor migration codebase - I will refactor soon but I wanted to just lift the existing tests while converting to anchor

File: programs/hyperplane/src/curve/stable.rs (hunk 1 above)
Comment (elliotkennedy) [reply programs/hyperplane/src/curve/stable.rs]: Thanks for your suggestions - I have implemented all apart from the `U128` which I will leave for now but change everywhere in the future. 

The divide by 0 is handled with `if factor > 1` also.

File: programs/hyperplane/src/curve/stable.rs [hunk 7]
```diff
@@ -216,14 +216,112 @@ fn compute_y(ann: u64, x: u128, d: u128) -> Result<u128> {
...
+        // Was there any remainder?
+        if round_up && factor * amount < source_amount {
+            amount + 1
+        } else {
+            amount
+        }
+    } else {
+        source_amount
+    };
+    Ok(amount)
```
Comment (oeble) [inline programs/hyperplane/src/curve/stable.rs:241]: `factor == 0` is not handled and I think it should return an error. At the moment, you handle it as if `factor == 1`, and this is not right.

**Review by oeble — APPROVED** [review]
LGTM!

File: programs/hyperplane/src/curve/stable.rs (hunk 7 above)
Comment (elliotkennedy) [reply programs/hyperplane/src/curve/stable.rs:246]: Added require_msg! on factor == 0

**Comment by elliotkennedy** [comment]
> LGTM!

I will merge this - checks haven't run for some reason

File: programs/hyperplane/sim/src/lib.rs [hunk 8]
```diff
@@ -111,7 +114,7 @@ impl StableSwapModel {
         self.balances[j as usize] =
             (y + &fee) * BigInt::from(10).pow(18) / &self.target_prices[i as usize];
 
-        (&dy - &fee).to_u128().unwrap()
+        ((&dy - &fee) / &self.rates[j as usize]).to_u128().unwrap()
     }
 
     pub fn sim_y(&self, i: u128, j: u128, x: u128) -> u128 {
```
Comment (andreihrs) [inline programs/hyperplane/sim/src/lib.rs:120]: you're not using the x param anywhere

File: programs/hyperplane/src/curve/calculator.rs [hunk 9]
```diff
@@ -98,12 +98,18 @@ pub trait CurveCalculator: Debug + DynAccountSerialize {
...
+    /// * `pool_token_a_amount` - the amount of token A in the pool
+    /// * `pool_token_b_amount` - the amount of token B in the pool
+    /// * `round_direction` - the direction to round the output trading token amounts
     fn pool_tokens_to_trading_tokens(
         &self,
         pool_tokens: u128,
         pool_token_supply: u128,
-        swap_token_a_amount: u128,
-        swap_token_b_amount: u128,
+        pool_token_a_amount: u128,
```
Comment (andreihrs) [inline programs/hyperplane/src/curve/calculator.rs:111]: think this doesn't respect the same convention you had for the other curves, so I'd think you'd either have to rename all of the others or revert this change

File: programs/hyperplane/sim/src/lib.rs (hunk 8 above)
Comment (elliotkennedy) [reply programs/hyperplane/sim/src/lib.rs:120]: Yes I am - it overwrites the value of x from `sim_xp` on line 123 (scaling comment is incorrect and will update)
